library/psa_crypto.c

```diff
@@ -567,9 +567,10 @@ static psa_status_t psa_load_rsa_representation( const psa_key_slot_t *slot,
     if( status != PSA_SUCCESS )
         goto exit;
 
-    /* Copy the PK-contained RSA context to the one provided as function input */
-    status = mbedtls_to_psa_error(
-                mbedtls_rsa_copy( rsa, mbedtls_pk_rsa( ctx ) ) );
+    /* Copy out the pointer to the RSA context, and reset the PK context
+     * such that pk_free doesn't free the RSA context we just grabbed. */
+    *p_rsa = mbedtls_pk_rsa( ctx );
```

Comment:
I agree, not having to have two copies of the same RSA key on the heap is a major benefit.

Re. the other point: yes, I could just let the PK context go out of scope, but that'll break when someone changes the PK context struct and adds more dynamically allocated members... IMO, the `_free()` function exists for a reason, and should be used always to avoid dangling memory. So to avoid that from happening, I have to do one of three things:
1. reach into the struct and abuse knowing what's going on inside that free function to avoid freeing the RSA representation (current solution, though not very clean)
2. copy the RSA representation out (but that results in a double copy on the heap, which isn't exactly preferred)
3. add a separate `_free_without_representation()` function to the PK module (which is arguably out of scope for this PR)

If there's agreement to go for #3 I can do that, but I'm afraid it'll just stall the PR even more, and there's a bunch of dependant PR's being staged.

Reply:
I don't think we need to go for #3. I pointed out the current solution to that problem on development as an already approved solution. Yours seems indeed better to me. Let's stick with that one for the time being.

Reply:
Let's not extend the PK module. Users of the library should not rely on the layout of `mbedtls_pk_context` because we may change it at any time. It's ok if we make use of knowledge of `mbedtls_pk_context` inside the library module, because if we do change the representation (or more likely, change an invariant about the representation, such as how memory is managed), we can update the code that uses it. We just need to do this sparingly.

I don't remember exactly where but there are a couple of precedents about assuming how pk contexts work. That's ok as long as the code is properly documented. In other words, I'm fine with #1 when cleaner ways are difficult or wasteful.

Reply:
OK, #1 it is.